### PR TITLE
hrt : display human readable rank, fix rank calculation, other minor changes

### PR DESCRIPTION
**Pull request by wonderingabout**
@roy7 @Dorus 

general look here : https://github.com/wonderingabout/gtp2ogs/tree/hrr-display-rank--fix-rank-calculation

many changes here, most important are :
- now displays rank in reject message in human readable rank (hrr) : for example 5d, 6k, 15k, etc...

- fixed the ranking calculation error : before that fix, a 5.3k is displayed as 5k on OGS, and a 4.8d is displayed as a 5d, while they are respectively 6k and 4d on OGS. Using Math.floor instead of to.Fixed()

- added warning if --maxmaintime is used in combination with --maxmaintimeranked or/and --maxmaintimeunranked

- added clearer reject message instructions if you use fischer time (need to change initial and max time for main time, then need to change increment time for period time. 

**TODO : maybe add a if rule==fischer, display another reject message using increment time**

- other minor changes

all tested to work successfully on beta : 

missioncomplete2
missioncomplete
log22
log11

@windo 

i saw that you rebased today again, maybe @roy7 wants to merge yours first if its possible
else, sorry again lol :+1:

### CODE REVIEW

File: gtp2ogs.js [hunk 1]
```diff
@@ -2226,9 +2248,9 @@ function num2gtpchar(num) { /* {{{ */
     return "abcdefghjklmnopqrstuvwxyz"[num];
 } /* }}} */
 function rank2str(r) { /* {{{ */
```
Comment (Dorus) [inline gtp2ogs.js]: I suggest we name this RankToString

File: gtp2ogs.js [hunk 2]
```diff
@@ -430,6 +430,22 @@ if (argv.id) {
     console.log("Warning: --id alias is no longer supported. Use --username instead.");
 }
 
+if (argv.maxmaintime && (argv.maxmaintimeranked || argv.maxmaintimeunranked)) {
+    console.log("Warning: You are using --maxmaintime in combination with --maxmaintimeranked or/and --maxmaintimeunranked. \n Use either --maxmaintime alone , OR --maxmaintimeranked with --maxmaintimeunranked. \n But don't use the 3 --maxmaintime arguments at the same time. ");
+}
+
+if (argv.minmaintime && (argv.minmaintimeranked || argv.minmaintimeunranked)) {
+    console.log("Warning: You are using --minmaintime in combination with --minmaintimeranked or/and --minmaintimeunranked. \n Use either --minmaintime alone , OR --minmaintimeranked with --minmaintimeunranked. \n But don't use the 3 --minmaintime arguments at the same time. ");
```
Comment (Dorus) [inline gtp2ogs.js]: it should be "and/or" not "or/and".

File: gtp2ogs.js [hunk 3]
```diff
@@ -1596,13 +1614,17 @@ class Connection {
         }
 
         if (user.ranking < argv.minrank) {
-            conn_log(user.username + " ranking too low: " + user.ranking);
-            return { reject: true, msg: "Minimum rank allowed is " + argv.minrank + " (in bot ranking units), your rank is " + user.ranking + " (in bot ranking units), your rank is too low, try again when your rank is high enough." };
+            let humanReadableUserRank = rank2str(user.ranking);
+            let humanReadableMinRank = rank2str(argv.minrank);
+            conn_log(user.username + " ranking too low: " + humanReadableUserRank + " : min is " + humanReadableMinRank);
+            return { reject: true, msg: "Minimum rank is " + humanReadableMinRank + " , your rank is too low, try again when your rank is high enough." };
```
Comment (Dorus) [inline gtp2ogs.js]: I think you should remove the space before `: min is` and `, your`.

File: gtp2ogs.js (hunk 2 above)
Comment (wonderingabout) [reply gtp2ogs.js]: i thought of this first, but one person is more likely to use 2 rather than 3 "family" arguments

File: gtp2ogs.js (hunk 3 above)
Comment (wonderingabout) [reply gtp2ogs.js]: the last space is fine i think
the first space maybe, but then i like spaces :laughing:

File: gtp2ogs.js (hunk 1 above)
Comment (wonderingabout) [reply gtp2ogs.js]: yes we can do it, but i dont want to rebase endlessly lol
so i can commit from website if needed, once all changes have been discussed :+1:

**Comment by wonderingabout** [comment]
also : 

TODO : maybe add a if rule==fischer, display another reject message using increment time

maybe i should do it on this PR

File: gtp2ogs.js (hunk 2 above)
Comment (Dorus) [reply gtp2ogs.js]: I'm just commenting on the grammar. In english you write "and/or". You never write "or/and".

**Comment by wonderingabout** [comment]
dont merge yet, i have a few other things to add on top of that, first thing being i found t.time_control is timecontrol name, and values are same as below : 

```
Feb 02 20:55:24   #  Bot is user id: 787
Feb 02 20:55:24   #  no autohandicap, rejecting challengefischer
Feb 02 20:55:24   #  Rejecting challenge from metaphysician splurgist :) (13k)  [19x19]  id = 4549
Feb 02 20:55:36   #  no autohandicap, rejecting challengesimple
Feb 02 20:55:36   #  Rejecting challenge from metaphysician splurgist :) (13k)  [19x19]  id = 4550
Feb 02 20:56:02   #  no autohandicap, rejecting challengecanadian
Feb 02 20:56:02   #  Rejecting challenge from metaphysician splurgist :) (13k)  [19x19]  id = 4551
Feb 02 20:56:10   #  no autohandicap, rejecting challengeabsolute
Feb 02 20:56:10   #  Rejecting challenge from metaphysician splurgist :) (13k)  [19x19]  id = 4552

```

**Comment by wonderingabout** [comment]
also maybe i'll try again to implement this 

i'm currently writing this : 

```
        if (argv.minmaintime || argv.minmaintimeranked || argv.minmaintimeunranked || argv.maxmaintime || argv.maxmaintimeranked || argv.maxmaintimeunranked) {
            if ( ["simple","none"].indexOf(t.time_control) >= 0) {
                conn_log("Minimum and Maximum main time is not supported in time control: " + t.time_control);
                return { reject: true, msg: "Minimum and Maximum main time is not supported in the time control " + t.time_control + ", please choose a time control that supports the use of a minimum and a maximum main time, such as byoyomi,fischer,canadian." };
            }            
            if (argv.minmaintime || argv.minmaintimeranked || argv.minmaintimeunranked) {
                let universalMinmaintimeValue = 0;
                let universalMinmaintimeConnSentence = user.username + " wanted main time below minimum main time ";
                let universalMinmaintimeBeginningSentence = "";
                let universalMinmaintimeEndingSentence = ", please increase main time \n - If you use Fischer, you need to change both -initial time- and -max time-, and also -time increment- \n - If you use canadian byo-yomi, set the average time per stone, for example 5 minutes/25 stones needs to be set up as 180 seconds (5 minutes) divided by 25 stones , which equals 7.2 seconds per stone. ";
                if (argv.minmaintime) {
                    universalMinmaintimeValue = timespanToDisplayString(argv.minmaintime);
                    universalMinmaintimeBeginningSentence = "Minimum main time is ";
                }
                if (argv.minmaintimeranked && notification.ranked) {
                    universalMinmaintimeValue = timespanToDisplayString(argv.minmaintimeranked);
                    universalMinmaintimeBeginningSentence = "Minimum main time for ranked games is ";
                }
                if (argv.minmaintimeunranked && !notification.ranked) {
                    universalMinmaintimeValue = timespanToDisplayString(argv.minmaintimeunranked);
                    universalMinmaintimeBeginningSentence = "Minimum main time for unranked games is ";
                }
                if (t.total_time   < universalMinmaintimeValue  || // absolute
                    t.initial_time < universalMinmaintimeValue  || // fischer
                    t.max_time     < universalMinmaintimeValue  || // fischer
                    t.main_time    < universalMinmaintimeValue) {  // others
                        conn_log(universalMinmaintimeConnSentence + universalMinmaintimeValue + ", rejecting challenge";
                        return { reject : true, msg: universalMinmaintimeBeginningSentence + universalMinmaintimeValue + universalMinmaintimeEndingSentence};
                }

            if (argv.maxmaintime || argv.maxmaintimeranked || argv.maxmaintimeunranked) {
                let universalMaxmaintimeValue = 0;
                let universalMaxmaintimeConnSentence = user.username + " wanted main time above maximum main time ";
                let universalMaxmaintimeBeginningSentence = "";
                let universalMaxmaintimeEndingSentence = ", please reduce main time \n - If you use Fischer, you need to change both -initial time- and -max time-, and also -time increment- \n - If you use canadian byo-yomi, set the average time per stone, for example 5 minutes/25 stones needs to be set up as 180 seconds (5 minutes) divided by 25 stones , which equals 7.2 seconds per stone. ";
                if (argv.maxmaintime) {
                    universalMaxmaintimeValue = timespanToDisplayString(argv.maxmaintime);
                    universalMaxmaintimeBeginningSentence = "Maximum main time is ";
                }
                if (argv.maxmaintimeranked && notification.ranked) {
                    universalMaxmaintimeValue = timespanToDisplayString(argv.maxmaintimeranked);
                    universalMaxmaintimeBeginningSentence = "Maximum main time for ranked games is ";
                } 
                if (argv.maxmaintimeunranked) {
                    universalMaxmaintimeValue = timespanToDisplayString(argv.maxmaintimeunranked);
                    universalMaxmaintimeBeginningSentence = "Maximum main time for unranked games is ";
                }
                if (t.total_time   < universalMaxmaintimeValue  || // absolute
                    t.initial_time < universalMaxmaintimeValue  || // fischer
                    t.max_time     < universalMaxmaintimeValue  || // fischer
                    t.main_time    < universalMaxmaintimeValue) {  // others
                        conn_log(universalMaxmaintimeConnSentence + universalMaxmaintimeValue + ", rejecting challenge";
                        return { reject : true, msg: universalMaxmaintimeBeginningSentence + universalMaxmaintimeValue + universalMaxmaintimeEndingSentence};
                }
        }

```

File: gtp2ogs.js (hunk 2 above)
Comment (roy7) [reply gtp2ogs.js]: Yeah it should be and/or

File: gtp2ogs.js (hunk 3 above)
Comment (roy7) [reply gtp2ogs.js]: Could you remove both spaces please? :)

File: gtp2ogs.js (hunk 1 above)
Comment (roy7) [reply gtp2ogs.js]: A minor change like renaming a function doesn't need rebasing. If you make the change to your branch and push the commit it'll automatically update the PR. Unless I don't understand what you meant.

Comment (wonderingabout) [reply gtp2ogs.js]: @roy7 

i meant i dont want to commit until i made all wanted changes, if i commit change function name then later there are other things to commit then i'll want to rebase, its what i meant :)

**Comment by wonderingabout** [comment]
this is now working, and it is very awesome !!!

will PR soon !!

**Comment by wonderingabout** [comment]
YAAAY it works !!!!!!!
@roy7 @windo @Dorus 
```
        ////// begining of *** UHMAEAT : Universal Highly Modulable And Expandable Argv Tree ***

        if (argv.minmaintime || argv.minmaintimeranked || argv.minmaintimeunranked || argv.maxmaintime || argv.maxmaintimeranked || argv.maxmaintimeunranked) {
            // later the t.time_control can't be used for rule detection for some reason,
            // so storing it now in a string while we can
            // also, whenever before TimecontrolString is going to be tested,
            // we always make sure it has the latest value
            // this avoids TimecontrolString being frozen on the same value independently from what user chooses, 
            // e.g. stuck on "absolute"
            // ** first we eliminate absolute and simple time control : they don't support the use of a period time
            // using absolute or simple (no period time) is likely to make bots timeout or play way too fast
            let universalMaintimeTimecontrolString = String(t.time_control);/*
            "fischer" , "simple", "byoyomi" , "canadian" , "absolute"*/
            if ((universalMaintimeTimecontrolString === "absolute") || (universalMaintimeTimecontrolString === "simple")) {
                conn_log(`Minimum and Maximum main time is not supported in time control: ${universalMaintimeTimecontrolString}` );
                return { reject: true, msg: `Period time management is not supported in the time control ${universalMaintimeTimecontrolString} , please choose a time control that supports the use of a period time, such as byoyomi,fischer,canadian.`};
            }          

            // ** then for fischer, byoyomi, or canadian, we use our UHMAEAT !
            let universalMaintimeMinimumMaximumSentence = "";    // minimum
            let universalMaintimeTimecontrolSentence = "";       // main time - initial time and/or max time, etc..
            let universalMaintimeForRankedUnrankedSentence = ""; // +/- for ranked/unranked games is
            let universalMaintimeNumber = 0;                     // for example 600 (600 seconds)
            let universalMaintimeToString = "";                  // for example "10 minutes"  = timespanToDisplayString(argv.xxx)
            let universalMaintimeIncreaseDecreaseSentence = "";  // , please increase/decrease
                                                                 // main time - MaintimeTimecontrolSentence again
            let universalMaintimeEndingSentence = "";            // optionnal, for example in canadian, adds explanation

            if (argv.minmaintime || argv.minmaintimeranked || argv.minmaintimeunranked) {
                universalMaintimeMinimumMaximumSentence = "Minimum ";
                universalMaintimeIncreaseDecreaseSentence = ", please increase ";
                let universalMaintimeConnSentence = user.username + " wanted main time below minimum main time ";
                if (argv.minmaintime) {
                    universalMaintimeNumber = argv.minmaintime;
                    universalMaintimeToString = timespanToDisplayString(argv.minmaintime);
                    universalMaintimeForRankedUnrankedSentence = "is ";
                }
                if (argv.minmaintimeranked && notification.ranked) {
                    universalMaintimeNumber = argv.minmaintimeranked;
                    universalMaintimeToString = timespanToDisplayString(argv.minmaintimeranked);
                    universalMaintimeForRankedUnrankedSentence = "for ranked games is ";
                }
                if (argv.minmaintimeunranked && !notification.ranked) {
                    universalMaintimeNumber = argv.minmaintimeunranked;
                    universalMaintimeToString = timespanToDisplayString(argv.minmaintimeunranked);
                    universalMaintimeForRankedUnrankedSentence = "for unranked games is ";
                 }
                // now just before TimecontrolString is being tested, we again make sure it has the latest value
                universalMaintimeTimecontrolString = String(t.time_control);/*
                "fischer" , "simple", "byoyomi" , "canadian" , "absolute"*/
                conn_log(universalMaintimeConnSentence + universalMaintimeToString + " in " + universalMaintimeTimecontrolString);
                if ((universalMaintimeTimecontrolString = "fischer") && ((t.initial_time < universalMaintimeNumber) || (t.max_time < universalMaintimeNumber))) {
                    universalMaintimeTimecontrolSentence = "Initial Time and/or Max Time ";
                    universalMaintimeEndingSentence = ".";
                } else if (universalMaintimeTimecontrolString === "byoyomi" && t.main_time < universalMaintimeNumber) {
                    universalMaintimeTimecontrolSentence = "Main Time ";
                    universalMaintimeEndingSentence = ".";
                } else if (universalMaintimeTimecontrolString === "canadian" && t.main_time < universalMaintimeNumber) {
                    universalMaintimeTimecontrolSentence = "Main Time ";
                    universalMaintimeEndingSentence = ". \n For canadian byo-yomi, set the average time per stone, for example 5 minutes/25 stones needs to be set up as 180 seconds (5 minutes) divided by 25 stones , which equals 7.2 seconds per stone.";
                // we comment out absolute and simple time since we dont want them to be supported
                // } else if (universalMaintimeTimecontrolString === "absolute" && t.total_time < universalMaintimeNumber) {
                //     universalMaintimeTimecontrolSentence = "Total Time ";
                //     universalMaintimeEndingSentence = ".";
                // } else if (universalMaintimeTimecontrolString === "simple" && t.main_time < universalMaintimeNumber) {
                //     universalMaintimeTimecontrolSentence = "Time per move ";
                //     universalMaintimeEndingSentence = ".";
                }
                return { reject : true, msg: universalMaintimeMinimumMaximumSentence + universalMaintimeTimecontrolSentence + universalMaintimeForRankedUnrankedSentence + universalMaintimeToString + universalMaintimeIncreaseDecreaseSentence + universalMaintimeTimecontrolSentence + universalMaintimeEndingSentence};
            }

            if (argv.maxmaintime || argv.maxmaintimeranked || argv.maxmaintimeunranked) {
                universalMaintimeMinimumMaximumSentence = "Maximum ";
                universalMaintimeIncreaseDecreaseSentence = ", please reduce ";
                let universalMaintimeConnSentence = user.username + " wanted main time above maximum main time ";
                if (argv.maxmaintime) {
                    universalMaintimeNumber = argv.maxmaintime;
                    universalMaintimeToString = timespanToDisplayString(argv.maxmaintime);
                    universalMaintimeForRankedUnrankedSentence = "is ";
                }
                if (argv.maxmaintimeranked && notification.ranked) {
                    universalMaintimeNumber = argv.maxmaintimeranked;
                    universalMaintimeToString = timespanToDisplayString(argv.maxmaintimeranked);
                    universalMaintimeForRankedUnrankedSentence = "for ranked games is ";
                }
                if (argv.maxmaintimeunranked && !notification.ranked) {
                    universalMaintimeNumber = argv.maxmaintimeunranked;
                    universalMaintimeToString = timespanToDisplayString(argv.maxmaintimeunranked);
                    universalMaintimeForRankedUnrankedSentence = "for unranked games is ";
                }
                // now just before TimecontrolString is being tested, we again make sure it has the latest value
                let universalMaintimeTimecontrolString = String(t.time_control);/*
                "fischer" , "simple", "byoyomi" , "canadian" , "absolute"*/
                conn_log(universalMaintimeConnSentence + universalMaintimeToString + " in " + universalMaintimeTimecontrolString);
                
                if ((universalMaintimeTimecontrolString === "fischer") && ((t.initial_time > universalMaintimeNumber) || (t.max_time > universalMaintimeNumber))) {
                    universalMaintimeTimecontrolSentence = "Initial Time and/or Max Time ";
                    universalMaintimeEndingSentence = ".";
                } else if (universalMaintimeTimecontrolString === "byoyomi" && t.main_time > universalMaintimeNumber) {
                    universalMaintimeTimecontrolSentence = "Main Time ";
                    universalMaintimeEndingSentence = ".";
                } else if (universalMaintimeTimecontrolString === "canadian" && t.main_time > universalMaintimeNumber) {
                    universalMaintimeTimecontrolSentence = "Main Time ";
                    universalMaintimeEndingSentence = ". \n For canadian byo-yomi, set the average time per stone, for example 5 minutes/25 stones needs to be set up as 180 seconds (5 minutes) divided by 25 stones , which equals 7.2 seconds per stone.";
                // we comment out absolute and simple time since we dont want them to be supported
                // } else if (universalMaintimeTimecontrolString === "absolute" && t.total_time > universalMaintimeNumber) {
                //     universalMaintimeTimecontrolSentence = "Total Time ";
                //     universalMaintimeEndingSentence = ".";
                // } else if (universalMaintimeTimecontrolString === "simple" && t.main_time > universalMaintimeNumber) {
                //     universalMaintimeTimecontrolSentence = "Time per move ";
                //     universalMaintimeEndingSentence = ".";
                }
                return { reject : true, msg: universalMaintimeMinimumMaximumSentence + universalMaintimeTimecontrolSentence + universalMaintimeForRankedUnrankedSentence + universalMaintimeToString + universalMaintimeIncreaseDecreaseSentence + universalMaintimeTimecontrolSentence + universalMaintimeEndingSentence};
            }
        }

        ////// end of *** UHMAEAT : Universal Highly Modulable And Expandable Argv Tree ***
```

screenshots below now :

**Comment by wonderingabout** [comment]
still one issue to solve though : 

when maintime is < argv.maxmaintime (e.g 10 minutes), challenge still gets rejected

**Comment by roy7** [comment]
Rebase now that #99 is in please, and I'll get it all merged.

**Comment by wonderingabout** [comment]
before rebasing i still want to add the UHMAEAT for number of periods example (" for example 5x30 seconds" if timecontrol != fischer)

**Comment by wonderingabout** [comment]
canadian messages were really really hard, but they are niice !!!


was using `--0ptr 30 --1ptr 60`

workssssssss
worksssssssssss

File: gtp2ogs.js [hunk 4]
```diff
@@ -418,34 +418,84 @@ process.on('uncaughtException', function (er) {
...
 })
 
+// console messages : for bot admins information
+
+console.log("You are using gtp2ogs version 6.0 \n - For changelog or devel updates, please visit : https://github.com/online-go/gtp2ogs/tree/devel \n - For known issues, please visit https://github.com/online-go/gtp2ogs/KNOWN-ISSUES.md");
+
+// console warnings : for bot admins usage
+
+if (argv.maxhandicap && (argv.maxhandicapranked || argv.maxhandicapunranked)) {
+    console.log("Warning: You are using --maxhandicap in combination with --maxhandicapranked or/and --maxhandicapunranked. \n Use either --maxhandicap alone , OR --maxhandicapranked with --maxhandicapunranked. \n But don't use the 3 --maxhandicap arguments at the same time. " );
```
Comment (Dorus) [inline gtp2ogs.js]: Please change this to "and/or", as that's correct english.

File: gtp2ogs.js [hunk 5]
```diff
@@ -418,34 +418,84 @@ process.on('uncaughtException', function (er) {
   }
 })
 
+// console messages : for bot admins information
+
+console.log("You are using gtp2ogs version 6.0 \n - For changelog or devel updates, please visit : https://github.com/online-go/gtp2ogs/tree/devel \n - For known issues, please visit https://github.com/online-go/gtp2ogs/KNOWN-ISSUES.md");
```
Comment (Dorus) [inline gtp2ogs.js]: Remove the extra space after "visit" in `please visit :`

File: gtp2ogs.js (hunk 4 above)
Comment (Dorus) [reply gtp2ogs.js]: Also remove the extra space before `\n`, that extra space risks console to turn over at the space, resulting in a double newline.

File: gtp2ogs.js [hunk 6]
```diff
@@ -418,34 +418,84 @@ process.on('uncaughtException', function (er) {
...
+
+// console warnings : for bot admins usage
+
+if (argv.maxhandicap && (argv.maxhandicapranked || argv.maxhandicapunranked)) {
+    console.log("Warning: You are using --maxhandicap in combination with --maxhandicapranked or/and --maxhandicapunranked. \n Use either --maxhandicap alone , OR --maxhandicapranked with --maxhandicapunranked. \n But don't use the 3 --maxhandicap arguments at the same time. " );
+}
+
+if (argv.minhandicap && (argv.minhandicapranked || argv.minhandicapunranked)) {
+    console.log("Warning: You are using --minhandicap in combination with --minhandicapranked or/and --minhandicapunranked. \n Use either --minhandicap alone , OR --minhandicapranked with --minhandicapunranked. \n But don't use the 3 --minhandicap arguments at the same time. " );
```
Comment (Dorus) [inline gtp2ogs.js]: Again so many spaces that needs to be removed. I'm going to name them again, but please fix them everywhere.
`or/and` => `and/or`
`. \n Use` => `.\nUse`
`alone ,` => `alone,`
`. \n But` => `.\nBut`
`time. " );` => `time.");`

File: gtp2ogs.js [hunk 7]
```diff
@@ -418,34 +418,84 @@ process.on('uncaughtException', function (er) {
...
+
+if (argv.nopause && (argv.nopauseranked || argv.nopauseunranked)) {
+    console.log("Warning: You are using --nopause in combination with --nopauseranked or/and --nopauseunranked. \n Use either --nopause alone , OR --nopauseranked with --nopauseunranked. \n But don't use the 3 --nopause arguments at the same time. " );
+}
+
+// console warnings : depreciated features if used
+
 if (argv.botid) {
-    console.log("Warning: --botid alias is no longer supported. Use --username instead.");
+    console.log("Warning: --botid alias is no longer supported. Use --username instead." );
```
Comment (Dorus) [inline gtp2ogs.js]: Stop adding random spaces at the end of the line. `" );` => `" );`. This space was not there in the original code.

File: gtp2ogs.js [hunk 8]
```diff
@@ -418,34 +418,84 @@ process.on('uncaughtException', function (er) {
...
+    console.log("Warning: --minrankedhandicap argument is no longer supported. Use --minhandicapunranked instead." );
 }
 
 if (argv.maxunrankedhandicap) {
-    console.log("Warning: --minrankedhandicap argument is no longer supported. Use --maxhandicapunranked instead.");
+    console.log("Warning: --minrankedhandicap argument is no longer supported. Use --maxhandicapunranked instead." );
 }
 
+
+
```
Comment (Dorus) [inline gtp2ogs.js]: Unnecessarily newlines are added here. Please remove.

File: gtp2ogs.js [hunk 9]
```diff
@@ -2225,10 +2423,21 @@ function num2gtpchar(num) { /* {{{ */
...
 } /* }}} */
-function rank2str(r) { /* {{{ */
-    r = r.toFixed();
-    if (r >= 30)  return (r-30+1) + 'd';
-    else          return (30-r) + 'k';
+function rankToString(r) { /* {{{ */
+    r = Math.floor(r);
+    if (r >= 30)  return (r-30+1) + 'd'; // r>=30 : 1 dan or stronger
+    else          return (30-r) + 'k'; // r<30 : 1 kyu or weaker
+} /* }}} */
```
Comment (Dorus) [inline gtp2ogs.js]: Add a newline here. 

Also more general question, who started adding those random `/* }}} */` and what purpose does it serve? I think we can safely remove all of them right?

File: gtp2ogs.js [hunk 10]
```diff
@@ -2225,10 +2423,21 @@ function num2gtpchar(num) { /* {{{ */
...
+function timespanToDisplayString(timespan) {
+    let ss = timespan % 60;
+    let mm = Math.floor(timespan / 60 % 60);
+    let hh = Math.floor(timespan / (60*60) % 24);
+    let dd = Math.floor(timespan / (60*60*24));
+    let text = ["days", "hours", "minutes", "seconds"];
+    return [dd, hh, mm, ss]
+    .map((e, i) => e === 0 ? "" : `${e} ${text[i]}`)
+    .filter(e => e !== "")
+    .join(" ");
```
Comment (Dorus) [inline gtp2ogs.js]: I would prefer 2-4 extra spaces at the start of line 2438-2440 to indicate the `.map.filter.join` lines are part of the previous line.

File: gtp2ogs.js [hunk 11]
```diff
@@ -1619,20 +1673,21 @@ class Connection {
...
-                .filter(e => e !== "")
-                .join(" ");
+
+        // ** first we eliminate absolute and simple time control : they don't support the use of a period time
+        // using absolute or simple (no period time) is likely to make bots timeout or play way too fast 
+        /* "fischer" , "simple", "byoyomi" , "canadian" , "absolute"*/
+        let TimeControlString = String(t.time_control);
+        if (TimeControlString === "absolute") {
+            conn_log("Minimum and Maximum main time is not supported in time control " + TimeControlString);
+            return { reject: true, msg: "Period time management is not supported in the time control " + TimeControlString + " , please choose a time control that supports the use of a period time, such as byoyomi,fischer,canadian." };
```
Comment (Dorus) [inline gtp2ogs.js]: I have some doubts about outright rejecting `absolute`, but that could be something we fix in another commit.

NOTE: This only reject it if we're using min and max time here right? I cannot see that clearly in the PR in this github view.

File: gtp2ogs.js [hunk 12]
```diff
@@ -1619,20 +1673,21 @@ class Connection {
...
         }
     
+        if ((TimeControlString === "simple") || (TimeControlString === "simple")) {
```
Comment (Dorus) [inline gtp2ogs.js]: Why are we doing the same check twice here?

File: gtp2ogs.js [hunk 13]
```diff
@@ -1725,197 +1780,340 @@ class Connection {
...
-                    conn_log(user.username + " wanted main time unranked above maxmaintimeunranked " + botSecondsTime);
-                    return { reject: true, msg: "Maximum main time for unranked games is " + botSecondsTime + ", please reduce main time. \n - If you use canadian byo-yomi, set the average time per stone, for example 5 minutes/25 stones needs to be set up as 180 seconds (5 minutes) divided by 25 stones , which equals 7.2 seconds per stone. " };
-            }
-        }
+        ////// end of *** UHMAEAT : Universal Highly Modulable And Expandable Argv Tree ***
 
         if (argv.minperiods && (t.periods < argv.minperiods)) {
             conn_log(user.username + " wanted too few periods: " + t.periods);
-            return { reject: true, msg: "Minimum number of periods is " + argv.minperiods + " , please increase the number of periods, for example " + argv.minperiods + " x 30 seconds " };
+            return { reject: true, msg: "Minimum number of periods is " + argv.minperiods + " , please increase the number of periods" };
```
Comment (Dorus) [inline gtp2ogs.js]: More spaces to pay attention to here `" ,` => `",`

**Comment by wonderingabout** [comment]
splitting this PR into several branches, after rebasing same as #99

File: gtp2ogs.js (hunk 8 above)
Comment (wonderingabout) [reply gtp2ogs.js]: done :1st_place_medal:

File: gtp2ogs.js (hunk 7 above)
Comment (wonderingabout) [reply gtp2ogs.js]: done :1st_place_medal:

File: gtp2ogs.js (hunk 4 above)
Comment (wonderingabout) [reply gtp2ogs.js]: done :1st_place_medal:

File: gtp2ogs.js (hunk 5 above)
Comment (wonderingabout) [reply gtp2ogs.js]: done :1st_place_medal:

File: gtp2ogs.js (hunk 3 above)
Comment (wonderingabout) [reply gtp2ogs.js]: i'm not sure which to remove lol, its hard :rofl: so will leave as it is :rofl:

File: gtp2ogs.js (hunk 11 above)
Comment (wonderingabout) [reply gtp2ogs.js]: no, in the past absolute time was always rejected independently from anything (same as why all rules default to chinese on ogs)

what i added here is that i rejected simple time too due to its confusing behaviour : 

it has a maintime with `t.main_time`
but it also has a period time with `t.per_move`

so i dont know how it should be handled with the arguments, hence why i removed it until we clear that out

File: gtp2ogs.js (hunk 12 above)
Comment (wonderingabout) [reply gtp2ogs.js]: mistake, was meant to be absolute+simple

**Comment by wonderingabout** [comment]
first PR : 
rank calculation and display fix and enhance :1st_place_medal: 
https://github.com/online-go/gtp2ogs/pull/113

second PR : 
enhance console behaviour
https://github.com/online-go/gtp2ogs/pull/114

File: gtp2ogs.js (hunk 10 above)
Comment (wonderingabout) [reply gtp2ogs.js]: not sure where to put them, so leaving as it is :100:

File: gtp2ogs.js (hunk 9 above)
Comment (wonderingabout) [reply gtp2ogs.js]: loool not me but its quite cool, let's keep them :+1: 

for the jump line not sure where to put again, so leaving as it is too :+1:

File: gtp2ogs.js (hunk 13 above)
Comment (wonderingabout) [reply gtp2ogs.js]: ok but there are so much already... :sob:
